feat: WIP fast automata decision procedure for bitvectors

tobiasgrosser opened the pull request:
This is still work-in-progress, but the changes are contained to the `Experimental` folder and does not break other things.
Therefore, we're merging it to prevent it from bitrotting.

ChrisHughes24 committed:
start on porting circuit code

ChrisHughes24 committed:
more proofs about FSM

ChrisHughes24 committed:
more stuff

ChrisHughes24 committed:
more stuff

ChrisHughes24 committed:
stuff

ChrisHughes24 committed:
why is there a error about well-founded recursion?

ChrisHughes24 committed:
more stuff

ChrisHughes24 committed:
redefine carry

ChrisHughes24 committed:
finish proving decidability of FSM stuff

ChrisHughes24 committed:
move some lemmas

ChrisHughes24 committed:
fix

ChrisHughes24 committed:
finish up

ChrisHughes24 committed:
add to decide a bit

ChrisHughes24 committed:
add trace

ChrisHughes24 committed:
use timeit

tobiasgrosser committed:
Drop unnecessary whitespace change

tobiasgrosser committed:
Add Decide to CI

tobiasgrosser committed:
make it compile

tobiasgrosser committed:
Merge remote-tracking branch 'origin/main' into fast_decide_updatre

tobiasgrosser committed:
Fix most of the compilation

tobiasgrosser committed:
fix up to one file

tobiasgrosser committed:
Proof some more

tobiasgrosser committed:
chore: update to lean 4.10.0-rc1 (#426)

tobiasgrosser commented on .github/workflows/ci.yml:
```diff
@@ -34,7 +34,7 @@ jobs:
         run: |
           lake -R exe cache get # download cache of mathlib docs.
-          lake -R build SSA.Experimental.Bits.Decide
+          lake -R build SSA.Experimental.Bits.Decide SSA.Experoimental.Bits.Fast.Decide
```
```suggestion
          lake -R build SSA.Experimental.Bits.Decide SSA.Experimental.Bits.Fast.Decide
```

tobiasgrosser committed:
Update .github/workflows/ci.yml

tobiasgrosser committed:
Merge remote-tracking branch 'origin/main' into fast_decide_updatre

tobiasgrosser committed:
Merge remote-tracking branch 'origin/main' into fast_decide_updatre

tobiasgrosser committed:
reduce diff

tobiasgrosser committed:
Merge remote-tracking branch 'origin/main' into fast_decide_updatre

alexkeizer committed:
fix: card_varsFinset_assignVars_lt proof

tobiasgrosser committed:
chore: compile-time-check Bits examples (#429)

tobiasgrosser committed:
fix some broken proofs

alexkeizer committed:
documentation

alexkeizer committed:
Merge branch 'fast_decide_updatre' of github.com:opencompl/lean-mlir into fast_decide_updatre

alexkeizer committed:
WIP: infinite bitstreams

alexkeizer committed:
WIP

alexkeizer committed:
WIP

tobiasgrosser committed:
update to lean 4.10.0 rc2 (#444)

tobiasgrosser committed:
Merge remote-tracking branch 'origin/main' into fast_decide_updatre

tobiasgrosser committed:
More proof compile

tobiasgrosser commented:
@alexkeizer, I added some sorries to make this build clean. I suggest we merge this to keep this code up-to-date with lean updates.

alexkeizer committed:
extract separate BitStream type and namespace

alexkeizer committed:
Merge branch 'fast_decide_updatre' of github.com:opencompl/lean-mlir into fast_decide_updatre

AtticusKuhn committed:
Feat: Change Tactic Auto (#448)

Pair programming with @bollu. Made several changes.

1. Added formatting and indentation to `TacticAuto.lean`.
2. Changed the ordering of the `bool -> arithmetic` tactic, so that we
solve 1 more statement (went from 35 to 34).

---------

Co-authored-by: Tobias Grosser <tobias@grosser.es>
Co-authored-by: Atticus Kuhn <atticusmkuhn@gmail.com>

tobiasgrosser committed:
inspiration: two more theorem classes (#446)

Co-authored-by: Atticus Kuhn <atticusmkuhn@gmail.com>

alexkeizer commented:
Agreed. It seems I've somehow broken the build again, I'll have a look at fixing that and then I'll hit merge

alexkeizer committed:
eliminate sorries

tobiasgrosser committed:
chore: fix build

tobiasgrosser commented:
Merge to ensure that we do not regress further.

tobiasgrosser committed:
chore: fix mathlib to 4.10.0-rc2 (#449)

tobiasgrosser commented:
@alexkeizer, this seems to be still quite broken. Maybe it's worth looking at this tomorrow.

tobiasgrosser committed:
chore: update to nightly-testing-2024-07-09 (#450)

tobiasgrosser committed:
feat: add Lean-MLIR publication to GitHub page (#451)

AnotherAlexHere committed:
chore: def to lemma (#438)

AtticusKuhn committed:
Allow Numeric Block Ids (#454)

Reference from the MLIR standard:




https://mlir.llvm.org/docs/LangRef/#identifiers-and-keywords


```
// Identifiers
bare-id ::= (letter|[_]) (letter|digit|[_$.])*
bare-id-list ::= bare-id (`,` bare-id)*
value-id ::= `%` suffix-id
alias-name :: = bare-id
suffix-id ::= (digit+ | ((letter|id-punct) (letter|id-punct|digit)*))
symbol-ref-id ::= `@` (suffix-id | string-literal) (`::` symbol-ref-id)?
value-id-list ::= value-id (`,` value-id)*
// Uses of value, e.g. in an operand list to an operation.
value-use ::= value-id (`#` decimal-literal)?
value-use-list ::= value-use (`,` value-use)*
```


Before, out parser did not allow numeric block ids. This change allows
numeric block ids.

Co-authored-by: Atticus Kuhn <atticusmkuhn@gmail.com>

AtticusKuhn committed:
Bugfix: Allow Properties with Carets in MLIR syntax. (#455)

Allow properties with carets in LLVM.

This is a syntax which was supported by the MLIR specification
[Properties](https://mlir.llvm.org/docs/LangRef/#properties)

but was not supported by our code.

---------

Co-authored-by: Atticus Kuhn <atticusmkuhn@gmail.com>

tobiasgrosser committed:
chore: update to nightly-testing-2024-07-11 (#456)

tobiasgrosser committed:
chore: drop unnecessary heartbeat (#459)

tobiasgrosser committed:
fear: add more complete but untyped mlir datastructures (#458)

alexkeizer committed:
WIP stash

AnotherAlexHere committed:
feat: Adding HackersDelight file (#457)

tobiasgrosser committed:
feat: Add DC dialect with Branch operator. (#407)

CC @ymherklotz, thanks for the pairing, 'twas a fun one ˆ_ˆ

---------

Co-authored-by: Alex Keizer <alex@keizer.dev>
Co-authored-by: Yann Herklotz <yann@yannherklotz.com>

tobiasgrosser committed:
chore: remove unnecessary includes and open namespaces (#461)

tobiasgrosser committed:
chore: fix typo in readme (#462)

tobiasgrosser committed:
feat: add leansat (#463)

tobiasgrosser committed:
Fix some compilation issues

tobiasgrosser commented on SSA/Experimental/Bits/Propagate.lean:
```diff
@@ -187,17 +188,17 @@ lemma id_eq_propagate (x : ℕ → Bool) :
+  | n+2 => simp [propagate_succ]
 
 lemma and_eq_propagate (x y : ℕ → Bool) :
     andSeq x y = propagate Empty.elim
```
@alexkeizer, it seems `andSeq` and friends must be updated to bitstreams.

tobiasgrosser committed:
Merge remote-tracking branch 'origin/main' into fast_decide_updatre

tobiasgrosser commented:
I merged the original fast_decide support, such that the diff here is now a lot smaller.

alexkeizer committed:
Merge branch 'fast_decide_updatre' of github.com:opencompl/lean-mlir into fast_decide_updatre

alexkeizer committed:
WIP

alexkeizer committed:
fix: eval_eq_iff_xor_eq_zero

alexkeizer committed:
WIP

Commit:
Make it not break

AtticusKuhn commented:
What is lsSeq?

Commit:
Fix issues with building

Commit:
remove some `sorry`s from Propagate

AtticusKuhn commented:
Why is this PR on the merge queue? There are still `sorry`s in some of the files.

tobiasgrosser commented:
Unsure. I may have misclicked.

Commit:
remove some `sorry`s

Commit:
remove a sorry

bollu commented on SSA/Experimental/Bits/Fast/BitStream.lean:
```diff
@@ -0,0 +1,410 @@
+theorem bmod_eq_of_ge_and_le (z : Int) (m : Nat)
+    (lower_bound : m/2 ≤ z) (upper_bound : z < m/2) :
+    z.bmod m = z := by
+  omega
```
```suggestion
theorem bmod_eq_of_ge_and_le (z : Int) (m : Nat)
    (hlower_bound : m/2 ≤ z) (hupper_bound : z < m/2) :
    z.bmod m = z := by
  omega
```

bollu commented:
Go for it. I didn't read it carefully, I'd suggest just merging it and moving on.

AtticusKuhn committed:
Rename Variable

Rename variables in accordance with @bollu's suggestion.

Co-authored-by: Siddharth <siddu.druid@gmail.com>
